cmd/federation-controller/main.go

```diff
@@ -206,18 +217,49 @@ func main() {
 		}
 	}
 
-	mcpServer := adss.NewServer(
-		&adss.ServerOpts{Port: 15010, ServerID: "mcp"},
-		mcpPushRequests,
-		onNewMCPSubscription,
-		mcp.NewGatewayResourceGenerator(istioConfigFactory),
-		mcp.NewServiceEntryGenerator(istioConfigFactory),
-		mcp.NewWorkloadEntryGenerator(istioConfigFactory),
-		mcp.NewVirtualServiceResourceGenerator(istioConfigFactory),
-		mcp.NewDestinationRuleResourceGenerator(istioConfigFactory),
-	)
-	if err := mcpServer.Run(ctx); err != nil {
-		log.Fatalf("Error running XDS server: %v", err)
+	if cfg.ConfigMode == config.ConfigModeMCP {
+		mcpServer := adss.NewServer(
+			&adss.ServerOpts{Port: 15010, ServerID: "mcp"},
+			meshConfigPushRequests,
+			startFDSClient,
+			mcp.NewGatewayResourceGenerator(istioConfigFactory),
+			mcp.NewServiceEntryGenerator(istioConfigFactory),
+			mcp.NewWorkloadEntryGenerator(istioConfigFactory),
+			mcp.NewVirtualServiceResourceGenerator(istioConfigFactory),
+			mcp.NewDestinationRuleResourceGenerator(istioConfigFactory),
+		)
+		if err := mcpServer.Run(ctx); err != nil {
+			log.Fatalf("Error running XDS server: %v", err)
+		}
+	} else if cfg.ConfigMode == config.ConfigModeK8s {
+		istioClient, err := istiokube.NewClient(istiokube.NewClientConfigForRestConfig(kubeConfig), "")
+		if err != nil {
+			log.Fatalf("failed to create Istio client: %v", err)
+		}
+
+		rm := kube.NewReconcilerManager(
+			meshConfigPushRequests,
+			kube.NewGatewayResourceReconciler(istioClient, istioConfigFactory),
+			kube.NewServiceEntryReconciler(istioClient, istioConfigFactory),
+			kube.NewWorkloadEntryReconciler(istioClient, istioConfigFactory),
+			kube.NewVirtualServiceReconciler(istioClient, istioConfigFactory),
+			kube.NewDestinationRuleReconciler(istioClient, istioConfigFactory),
+		)
+		if err := rm.ReconcileAll(ctx); err != nil {
+			log.Fatalf("initial Istio resource reconciliation failed: %v", err)
+		}
+		go rm.Start(ctx)
+
+		// TODO: implement an appropriate wait allowing proxy to receive all initial
+		// federation config before starting FDS client.
+		time.Sleep(time.Second * 5)
```

Comment:
I do not think we need to wait, because the client will try to reconnect in case of failures.

Reply:
As far as I remember, the connection was never established without this wait.

Reply:
I think it failed [here](https://github.com/openshift-service-mesh/federation/blob/master/internal/pkg/xds/adsc/adsc.go#L70).
Maybe it would be possible to call `a.reconnect` when an error is returned here.